upf/problem.py

```diff
@@ -176,13 +197,88 @@ def set_initial_value(self, fluent: Union[FNode, upf.Fluent],
+        else:
             raise UPFProblemDefinitionError('Initial value not set!')
-        return self._initial_value[fluent_exp]
+
+    def domain_size(self, typename: upf.typing.Type) -> Optional[int]:
+        """Returns the domain size of the given type."""
+        if typename.is_bool_type():
+            return 2
+        elif typename.is_user_type():
+            return len(self.objects(typename))
+        elif typename.is_real_type():
+            lb = typename.lower_bound() # type: ignore
+            ub = typename.upper_bound() # type: ignore
+            if lb is not None and lb == ub:
+                return 1
+        elif typename.is_int_type():
+            lb = typename.lower_bound() # type: ignore
+            ub = typename.upper_bound() # type: ignore
+            if lb is not None and ub is not None:
+                return ub - lb
+        return None
+
+    def domain_item(self, typename: upf.typing.Type, idx: int) -> FNode:
+        """Returns the ith domain item of the given type."""
+        if typename.is_bool_type():
+            return self._env.expression_manager.Bool(idx == 0)
+        elif typename.is_user_type():
+            return self._env.expression_manager.ObjectExp(self.objects(typename)[idx])
+        elif typename.is_real_type():
+            lb = typename.lower_bound() # type: ignore
+            ub = typename.upper_bound() # type: ignore
+            assert (lb is not None and lb == ub)
+            return self._env.expression_manager.Real(lb)
+        elif typename.is_int_type():
+            lb = typename.lower_bound() # type: ignore
+            ub = typename.upper_bound() # type: ignore
+            assert (lb is not None and ub is not None)
+            return self._env.expression_manager.Int(lb + idx)
+        else:
+            raise
+
+    def get_ith_fluent_exp(self, fluent: upf.Fluent, domain_sizes: List[int], idx: int) -> FNode:
+        """Returns the ith ground fluent expression."""
+        quot = idx
+        rem = 0
+        actual_parameters = []
+        for i in range(fluent.arity()):
+            ds = domain_sizes[i];
+            rem = quot % ds
+            quot //= ds
+            v = self.domain_item(fluent.signature()[i], rem)
+            actual_parameters.append(v)
+        return fluent(*actual_parameters)
 
     def initial_values(self) -> Dict[FNode, FNode]:
         """Gets the initial value of the fluents."""
-        return self._initial_value
+        res = self._initial_value
+        for f in self._fluents.values():
+            if f.arity() == 0:
+                f_exp = self._env.expression_manager.FluentExp(f)
+                res[f_exp] = self.initial_value(f_exp)
+            else:
+                ground_size = 1
+                domain_sizes = []
+                is_groundable = True
+                for p in f.signature():
+                    ds = self.domain_size(p)
+                    if ds is None:
+                        is_groundable = False
```

Comment:
Maybe we can simply assume that all the parameters of all the fluents will be groundable. Otherwise, this way of defining initial values does not work. No?

Reply:
Also, the portions of the code dealing with INT and REAL parameters are currently totally untested, decreasing the coverage and complicating the code

Reply:
Do you mean to assume that all the fluents parameters have type UserType?

Reply:
I updated the code assuming groundable parameters.